packages/@uppy/form/src/index.ts

```diff
@@ -0,0 +1,174 @@
+
+  handleUploadStart(): void {
+    if (this.opts.getMetaFromForm) {
+      this.getMetaFromForm()
+    }
+  }
+
+  handleSuccess(result: Result<M, B>): void {
+    if (this.opts.addResultToForm) {
+      this.addResultToForm(result)
+    }
+
+    if (this.opts.submitOnSuccess) {
+      this.form.requestSubmit()
+    }
+  }
+
+  handleFormSubmit(ev: Event): void {
+    if (this.opts.triggerUploadOnSubmit) {
+      ev.preventDefault()
+      const elements = toArray((ev.target as HTMLFormElement).elements)
+      const disabledByUppy: HTMLButtonElement[] = []
+      elements.forEach((el) => {
+        const isButton =
+          el.tagName === 'BUTTON' ||
+          (el.tagName === 'INPUT' &&
+            (el as HTMLButtonElement).type === 'submit')
+        if (isButton && !(el as HTMLButtonElement).disabled) {
+          ;(el as HTMLButtonElement).disabled = true // eslint-disable-line no-param-reassign
+          disabledByUppy.push(el as HTMLButtonElement)
+        }
+      })
+      this.uppy
+        .upload()
+        .then(
+          () => {
+            disabledByUppy.forEach((button) => {
+              button.disabled = false // eslint-disable-line no-param-reassign
+            })
+          },
+          (err) => {
+            disabledByUppy.forEach((button) => {
+              button.disabled = false // eslint-disable-line no-param-reassign
+            })
+            return Promise.reject(err)
+          },
+        )
+        .catch((err) => {
+          this.uppy.log(err.stack || err.message || err)
+        })
+    }
+  }
+
+  addResultToForm(result: Result<M, B>): void {
+    this.uppy.log('[Form] Adding result to the original form:')
+    this.uppy.log(result)
+
+    let resultInput: HTMLInputElement | null = this.form.querySelector(
+      `[name="${this.opts.resultName}"]`,
+    )
+    if (resultInput) {
+      // Append new result to the previous result array.
+      // If the previous result is empty, or not an array,
+      // set it to an empty array.
+      let updatedResult
+      try {
+        updatedResult = JSON.parse(resultInput.value)
+      } catch (err) {
+        // Nothing, since we check for array below anyway
+      }
+
+      if (!Array.isArray(updatedResult)) {
+        updatedResult = []
+      }
+      updatedResult.push(result)
+      resultInput.value = JSON.stringify(updatedResult)
+      return
+    }
+
+    resultInput = document.createElement('input')
+    resultInput.name = this.opts.resultName
+    resultInput.type = 'hidden'
+    resultInput.value = JSON.stringify([result])
+
+    this.form.appendChild(resultInput)
+  }
+
+  getMetaFromForm(): void {
+    const formMeta = getFormData(this.form)
+    // We want to exclude meta the the Form plugin itself has added
+    // See https://github.com/transloadit/uppy/issues/1637
+    delete formMeta[this.opts.resultName]
+    this.uppy.setMeta(formMeta)
+  }
+
+  install(): void {
+    this.form = findDOMElement(this.opts.target) as HTMLFormElement
```

Comment:
Let me repeat what I said in https://github.com/transloadit/uppy/pull/4946#discussion_r1499429360 as it's relevant here: we might want to use a runtime assertion instead of a TS one:

```js
function assertHTMLFormElement(input: unknown): HTMLFormElement {
  if ((input as Node)?.nodeName !== 'FORM') {
    throw new Error('ASSERTION FAILED: the target is not a <form> element', {cause: input})
  }
  return input as any
}
```
```suggestion
    this.form = assertHTMLFormElement(findDOMElement(this.opts.target))
```

That might be a case where a runtime check can be more helpful for debugging than a TypeError at some random place. No strong feelings on my side, maybe @mifi have thoughts?

Reply:
Yes, I think that's better. Although we should probably `try { }` the assertHTMLFormElement call and catch the error and `this.uppy.log(err.message, 'error')` in order to not make a potentially breaking change (throwing a new error that we previously didnt throw)

Reply:
It's not a breaking change. If the element isn't found, than it would have crashed internally. With a new error it crashes all the same.